File: apps/ios/PagApp/PagApp/Services/SurveyService.swift
import Foundation
import Combine

@MainActor
public class SurveyService: ObservableObject {
    public static let shared = SurveyService()
    
    @Published public private(set) var eligibleSurveys: [PAGSurvey] = []
    @Published public private(set) var completedSurveys: [PAGSurvey] = []
    @Published public private(set) var isLoading: Bool = false
    @Published public private(set) var errorMessage: String? = nil
    
    private init() {}
    
    public func fetchEligibleSurveys() async {
        self.isLoading = true
        self.errorMessage = nil
        
        do {
            let response = try await PAGApiClient.shared.get(endpoint: "/home")
            if let success = response["success"] as? Bool, success,
               let data = response["data"] as? [String: Any],
               let rawSurveys = data["surveys"] as? [[String: Any]] {
                
                var parsedList: [PAGSurvey] = []
                for item in rawSurveys {
                    if let surveyId = item["id"] as? String ?? item["surveyId"] as? String,
                       let title = item["title"] as? String {
                        
                        var qList: [PAGQuestion] = []
                        if let rawQs = item["questions"] as? [[String: Any]] {
                            for (qIdx, q) in rawQs.enumerated() {
                                let qId = q["questionId"] as? String ?? "q\(qIdx + 1)"
                                let qText = q["text"] as? String ?? ""
                                var options: [PAGQuestionOption] = []
                                if let rawOpts = q["options"] as? [String] {
                                    for (oIdx, optText) in rawOpts.enumerated() {
                                        options.append(PAGQuestionOption(optionId: "opt_\(oIdx + 1)", label: optText, order: oIdx + 1))
                                    }
                                } else if let rawOptObjects = q["options"] as? [[String: Any]] {
                                    for (oIdx, optObj) in rawOptObjects.enumerated() {
                                        let oId = optObj["optionId"] as? String ?? optObj["id"] as? String ?? "opt_\(oIdx + 1)"
                                        let oLabel = optObj["label"] as? String ?? optObj["text"] as? String ?? "Seçenek \(oIdx + 1)"
                                        let oOrder = optObj["order"] as? Int ?? (oIdx + 1)
                                        options.append(PAGQuestionOption(optionId: oId, label: oLabel, order: oOrder))
                                    }
                                } else if let rawOptAny = q["options"] as? [Any] {
                                    for (oIdx, item) in rawOptAny.enumerated() {
                                        if let str = item as? String {
                                            options.append(PAGQuestionOption(optionId: "opt_\(oIdx + 1)", label: str, order: oIdx + 1))
                                        } else if let dict = item as? [String: Any] {
                                            let oId = dict["optionId"] as? String ?? dict["id"] as? String ?? "opt_\(oIdx + 1)"
                                            let oLabel = dict["label"] as? String ?? dict["text"] as? String ?? "Seçenek \(oIdx + 1)"
                                            let oOrder = dict["order"] as? Int ?? (oIdx + 1)
                                            options.append(PAGQuestionOption(optionId: oId, label: oLabel, order: oOrder))
                                        }
                                    }
                                }
                                
                                if options.isEmpty {
                                    options = [
                                        PAGQuestionOption(optionId: "opt_1", label: "Evet / Katılıyorum", order: 1),
                                        PAGQuestionOption(optionId: "opt_2", label: "Hayır / Katılmıyorum", order: 2)
                                    ]
                                }
                                
                                qList.append(PAGQuestion(questionId: qId, order: qIdx + 1, type: "SINGLE_SELECT", text: qText, options: options))
                            }
                        }
                        
                        let survey = PAGSurvey(
                            surveyId: surveyId,
                            ownerType: item["ownerType"] as? String ?? "PAG",
                            organizationId: item["organizationId"] as? String,
                            surveyType: item["surveyType"] as? String ?? "PAG",
                            title: title,
                            description: item["description"] as? String ?? "",
                            status: item["status"] as? String ?? "ACTIVE",
                            questionCount: qList.count > 0 ? qList.count : 3,
                            questions: qList,
                            profileScoreReward: item["profileScoreReward"] as? Int ?? 50,
                            isCompleted: false,
                            isHighlighted: item["isHighlighted"] as? Bool ?? false
                        )
                        parsedList.append(survey)
                    }
                }
                
                self.eligibleSurveys = parsedList
            } else {
                self.eligibleSurveys = []
            }
            self.isLoading = false
        } catch {
            print("[SurveyService] Fetch eligible surveys error: \(error.localizedDescription)")
            self.eligibleSurveys = []
            self.isLoading = false
        }
    }

    public func fetchCompletedSurveys() async {
        do {
            if let response = try? await PAGApiClient.shared.get(endpoint: "/surveys/completed"),
               let success = response["success"] as? Bool, success,
               let data = response["data"] as? [String: Any],
               let rawSurveys = data["completedSurveys"] as? [[String: Any]] {
                
                var parsedList: [PAGSurvey] = []
                for item in rawSurveys {
                    if let surveyId = item["surveyId"] as? String ?? item["id"] as? String,
                       let title = item["title"] as? String {
                        let survey = PAGSurvey(
                            surveyId: surveyId,
                            ownerType: item["ownerType"] as? String ?? "PAG",
                            organizationId: item["organizationId"] as? String,
                            surveyType: item["surveyType"] as? String ?? "PAG",
                            title: title,
                            description: item["description"] as? String ?? "",
                            status: "COMPLETED",
                            questionCount: item["questionCount"] as? Int ?? 3,
                            questions: [],
                            profileScoreReward: item["profileScoreReward"] as? Int ?? 50,
                            isCompleted: true,
                            isHighlighted: false
                        )
                        parsedList.append(survey)
                    }
                }
                self.completedSurveys = parsedList
            }
        } catch {
            print("[SurveyService] Fetch completed error: \(error.localizedDescription)")
        }
    }
    
    public func getSurveyById(surveyId: String) throws -> PAGSurvey {
        if let found = eligibleSurveys.first(where: { $0.surveyId == surveyId }) {
            return found
        }
        if let found = completedSurveys.first(where: { $0.surveyId == surveyId }) {
            return found
        }
        throw NSError(domain: "SurveyService", code: 404, userInfo: [NSLocalizedDescriptionKey: "Anket bulunamadı veya pasif durumda."])
    }
    
    public func fetchSurveyDetail(surveyId: String) async throws -> PAGSurvey {
        if let found = eligibleSurveys.first(where: { $0.surveyId == surveyId }) {
            return found
        }
        if let found = completedSurveys.first(where: { $0.surveyId == surveyId }) {
            return found
        }
        await fetchEligibleSurveys()
        return try getSurveyById(surveyId: surveyId)
    }
    
    public func submitSurveyResponse(surveyId: String, answers: [PAGAnswerInput], isProfile: Bool = false) async throws -> PAGSurveyCompletionResult {
        let answerDicts = answers.map { ["questionId": $0.questionId, "optionId": $0.optionId] }
        
        let apiResult = try await PAGApiClient.shared.post(
            endpoint: "/surveys/\(surveyId)/submit",
            body: ["answers": Array(answerDicts.prefix(3))]
        )
        
        guard let success = apiResult["success"] as? Bool, success,
              let dataDict = apiResult["data"] as? [String: Any] else {
            let errMsg = apiResult["error"] as? String ?? "Tamamlama işlemi başarısız."
            throw NSError(domain: "SurveyService", code: 400, userInfo: [NSLocalizedDescriptionKey: errMsg])
        }
        
        let scoreAwarded = dataDict["earnedScore"] as? Int ?? 50
        let currentScore = dataDict["profileScore"] as? Int
        let earnedReward = dataDict["earnedReward"] as? [String: Any]
        let prizeAmount = earnedReward?["amount"] as? Int
        let prizeType = earnedReward?["type"] as? String
        let vCode = earnedReward?["code"] as? String
        let vTitle = earnedReward?["poolName"] as? String
        let currentRewardBalance = Int(Double(dataDict["rewardBalance"] as? String ?? "0") ?? 0)
        
        await fetchEligibleSurveys()
        await UserService.shared.bootstrapCurrentUser()
        
        return PAGSurveyCompletionResult(
            responseId: "\(surveyId)_submitted",
            surveyId: surveyId,
            completedAt: ISO8601DateFormatter().string(from: Date()),
            isDuplicate: false,
            profileScorePotential: scoreAwarded,
            currentProfileScore: currentScore,
            rewardAwarded: prizeAmount,
            rewardType: prizeType,
            voucherCode: vCode,
            voucherTitle: vTitle,
            currentRewardBalance: currentRewardBalance
        )
    }
}

File: apps/android/app/src/main/java/com/alafteknoloji/pagapp/services/SurveyService.kt
package com.alafteknoloji.pagapp.services

import com.alafteknoloji.pagapp.models.PAGAnswerInput
import com.alafteknoloji.pagapp.models.PAGQuestion
import com.alafteknoloji.pagapp.models.PAGQuestionOption
import com.alafteknoloji.pagapp.models.PAGSurvey
import com.alafteknoloji.pagapp.models.PAGSurveyCompletionResult
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import org.json.JSONArray
import org.json.JSONObject

class SurveyService {

    private val _eligibleSurveys = MutableStateFlow<List<PAGSurvey>>(emptyList())
    val eligibleSurveys: StateFlow<List<PAGSurvey>> = _eligibleSurveys.asStateFlow()

    private val _completedSurveys = MutableStateFlow<List<PAGSurvey>>(emptyList())
    val completedSurveys: StateFlow<List<PAGSurvey>> = _completedSurveys.asStateFlow()

    private val _isLoading = MutableStateFlow(false)
    val isLoading: StateFlow<Boolean> = _isLoading.asStateFlow()

    private val _errorMessage = MutableStateFlow<String?>(null)
    val errorMessage: StateFlow<String?> = _errorMessage.asStateFlow()

    suspend fun fetchEligibleSurveys() {
        _isLoading.value = true
        _errorMessage.value = null

        try {
            val apiRes = PAGApiClient.get("/home")
            if (apiRes != null && apiRes.optBoolean("success")) {
                val dataObj = apiRes.optJSONObject("data")
                val rawSurveys = dataObj?.optJSONArray("surveys")
                val parsed = mutableListOf<PAGSurvey>()
                if (rawSurveys != null) {
                    for (i in 0 until rawSurveys.length()) {
                        val item = rawSurveys.getJSONObject(i)
                        val surveyId = item.optString("id").ifEmpty { item.optString("surveyId") }
                        val title = item.optString("title")
                        if (surveyId.isNotEmpty() && title.isNotEmpty()) {
                            val qList = mutableListOf<PAGQuestion>()
                            val rawQs = item.optJSONArray("questions")
                            if (rawQs != null) {
                                for (qIdx in 0 until rawQs.length()) {
                                    val qObj = rawQs.getJSONObject(qIdx)
                                    val qId = qObj.optString("questionId", "q${qIdx + 1}")
                                    val qText = qObj.optString("text", "")
                                    val optList = mutableListOf<PAGQuestionOption>()
                                    val rawOpts = qObj.optJSONArray("options")
                                    if (rawOpts != null) {
                                        for (oIdx in 0 until rawOpts.length()) {
                                            val opt = rawOpts.get(oIdx)
                                            if (opt is String) {
                                                optList.add(PAGQuestionOption("opt_${oIdx + 1}", opt, oIdx + 1))
                                            } else if (opt is JSONObject) {
                                                optList.add(PAGQuestionOption(
                                                    opt.optString("optionId", "opt_${oIdx + 1}"),
                                                    opt.optString("label", ""),
                                                    opt.optInt("order", oIdx + 1)
                                                ))
                                            }
                                        }
                                    }
                                    qList.add(PAGQuestion(qId, qIdx + 1, "SINGLE_SELECT", qText, optList))
                                }
                            }

                            parsed.add(
                                PAGSurvey(
                                    surveyId = surveyId,
                                    ownerType = item.optString("ownerType", "PAG"),
                                    organizationId = if (item.isNull("organizationId")) null else item.optString("organizationId"),
                                    surveyType = item.optString("surveyType", "PAG"),
                                    title = title,
                                    description = item.optString("description", ""),
                                    status = item.optString("status", "ACTIVE"),
                                    questionCount = if (qList.isNotEmpty()) qList.size else item.optInt("questionCount", 3),
                                    questions = qList,
                                    profileScoreReward = item.optInt("profileScoreReward", 50),
                                    isCompleted = false,
                                    isHighlighted = item.optBoolean("isHighlighted", false)
                                )
                            )
                        }
                    }
                }
                _eligibleSurveys.value = parsed
            } else {
                _eligibleSurveys.value = emptyList()
            }
        } catch (e: Exception) {
            e.printStackTrace()
            _eligibleSurveys.value = emptyList()
        } finally {
            _isLoading.value = false
        }
    }

    suspend fun fetchCompletedSurveys() {
        try {
            val apiRes = PAGApiClient.get("/surveys/completed")
            if (apiRes != null && apiRes.optBoolean("success")) {
                val dataDict = apiRes.optJSONObject("data")
                val rawSurveys = dataDict?.optJSONArray("completedSurveys")
                val parsed = mutableListOf<PAGSurvey>()
                if (rawSurveys != null) {
                    for (i in 0 until rawSurveys.length()) {
                        val item = rawSurveys.getJSONObject(i)
                        val surveyId = item.optString("surveyId").ifEmpty { item.optString("id") }
                        val title = item.optString("title")
                        if (surveyId.isNotEmpty() && title.isNotEmpty()) {
                            parsed.add(
                                PAGSurvey(
                                    surveyId = surveyId,
                                    ownerType = item.optString("ownerType", "PAG"),
                                    organizationId = if (item.isNull("organizationId")) null else item.optString("organizationId"),
                                    surveyType = item.optString("surveyType", "PAG"),
                                    title = title,
                                    description = item.optString("description", ""),
                                    status = "COMPLETED",
                                    questionCount = item.optInt("questionCount", 3),
                                    profileScoreReward = item.optInt("profileScoreReward", 50),
                                    isCompleted = true
                                )
                            )
                        }
                    }
                }
                _completedSurveys.value = parsed
            }
        } catch (e: Exception) {
            e.printStackTrace()
        }
    }

    suspend fun fetchSurveyDetail(surveyId: String): PAGSurvey? {
        val found = _eligibleSurveys.value.find { it.surveyId == surveyId }
        if (found != null) return found
        val comp = _completedSurveys.value.find { it.surveyId == surveyId }
        return comp
    }

    suspend fun submitSurveyResponse(
        surveyId: String,
        answers: List<PAGAnswerInput>,
        isProfile: Boolean = false,
        userService: UserService? = null
    ): PAGSurveyCompletionResult {
        val answersArray = JSONArray()
        answers.take(3).forEach { ans ->
            val obj = JSONObject().apply {
                put("questionId", ans.questionId)
                put("optionId", ans.optionId)
            }
            answersArray.put(obj)
        }

        val jsonBody = JSONObject().apply {
            put("answers", answersArray)
        }

        val apiRes = PAGApiClient.post("/surveys/$surveyId/submit", jsonBody)
        if (apiRes != null && apiRes.optBoolean("success")) {
            val dataDict = apiRes.optJSONObject("data")
            val scoreAwarded = dataDict?.optInt("earnedScore", 50) ?: 50
            val currentScore = dataDict?.optInt("profileScore")
            val earnedReward = dataDict?.optJSONObject("earnedReward")
            val prizeAmount = earnedReward?.optInt("amount")
            val prizeType = earnedReward?.optString("type")
            val vCode = earnedReward?.optString("code")
            val vTitle = earnedReward?.optString("poolName")
            val currentRewardBalance = dataDict?.optString("rewardBalance", "0")?.toDoubleOrNull()?.toInt() ?: 0

            fetchEligibleSurveys()
            userService?.bootstrapCurrentUser()

            return PAGSurveyCompletionResult(
                responseId = "${surveyId}_submitted",
                surveyId = surveyId,
                completedAt = java.text.SimpleDateFormat("yyyy-MM-dd'T'HH:mm:ss'Z'", java.util.Locale.US).format(java.util.Date()),
                isDuplicate = false,
                profileScorePotential = scoreAwarded,
                currentProfileScore = currentScore,
                rewardAwarded = prizeAmount,
                rewardType = prizeType,
                voucherCode = vCode,
                voucherTitle = vTitle,
                currentRewardBalance = currentRewardBalance
            )
        }

        val errorMsg = apiRes?.optString("error", "Tamamlama işlemi başarısız.") ?: "Tamamlama işlemi başarısız."
        throw Exception(errorMsg)
    }
}
